File: frontend/src/api/llmConfig.ts
import api from "./http";
import type { DiscoverModelsResponse, UserLLMConfig } from "../types";


export function getLLMConfig() {
  return api.get<UserLLMConfig>("/api/v1/llm-config");
}

export function updateLLMConfig(payload: {
  provider_name?: string;
  base_url?: string;
  default_model?: string;
  agent_model_overrides?: Record<string, string>;
  agent_overrides?: Record<string, {
    override_enabled: boolean;
    provider_name?: string;
    base_url?: string;
    default_model?: string;
    api_key?: string;
    clear_api_key?: boolean;
  }>;
  api_key?: string;
  enabled?: boolean;
  clear_api_key?: boolean;
}) {
  return api.put<UserLLMConfig>("/api/v1/llm-config", payload);
}

export function discoverLLMModels(payload: {
  agent_code?: string;
  provider_name?: string;
  base_url?: string;
  api_key?: string;
}) {
  return api.post<DiscoverModelsResponse>("/api/v1/llm-config/discover-models", payload);
}

File: backend/app/schemas/llm_config.py
import datetime as dt
from typing import Optional

from pydantic import BaseModel, Field


class AgentModelOption(BaseModel):
    agent_code: str
    display_name: str
    role_default_model: str
    enabled: bool
    override_enabled: bool = False
    provider_name: str = ""
    base_url: str = ""
    default_model: str = ""
    has_api_key: bool = False
    masked_api_key: str = ""
    is_ready: bool = False


class AgentRuntimeOverrideUpdate(BaseModel):
    override_enabled: bool = False
    provider_name: Optional[str] = Field(default=None, max_length=120)
    base_url: Optional[str] = Field(default=None, max_length=255)
    default_model: Optional[str] = Field(default=None, max_length=120)
    api_key: Optional[str] = Field(default=None, max_length=500)
    clear_api_key: bool = False


class UserLLMConfigUpdate(BaseModel):
    provider_name: Optional[str] = Field(default=None, max_length=120)
    base_url: Optional[str] = Field(default=None, max_length=255)
    default_model: Optional[str] = Field(default=None, max_length=120)
    agent_model_overrides: Optional[dict[str, str]] = None
    agent_overrides: Optional[dict[str, AgentRuntimeOverrideUpdate]] = None
    api_key: Optional[str] = Field(default=None, max_length=500)
    enabled: Optional[bool] = None
    clear_api_key: bool = False


class DiscoverModelsRequest(BaseModel):
    agent_code: Optional[str] = Field(default=None, max_length=120)
    provider_name: Optional[str] = Field(default=None, max_length=120)
    base_url: Optional[str] = Field(default=None, max_length=255)
    api_key: Optional[str] = Field(default=None, max_length=500)


class DiscoveredModelOption(BaseModel):
    model_id: str
    owned_by: Optional[str] = None


class DiscoverModelsResponse(BaseModel):
    provider_name: str
    base_url: str
    used_saved_api_key: bool
    models: list[DiscoveredModelOption] = Field(default_factory=list)


class UserLLMConfigResponse(BaseModel):
    provider_name: str
    base_url: str
    default_model: str
    enabled: bool
    has_api_key: bool
    masked_api_key: str
    agent_model_overrides: dict[str, str] = Field(default_factory=dict)
    available_roles: list[AgentModelOption] = Field(default_factory=list)
    is_ready: bool
    updated_at: Optional[dt.datetime] = None
